src/nafill.c

```diff
@@ -86,6 +86,28 @@ void nafillInteger64(int64_t *x, uint_fast64_t nx, unsigned int type, int64_t fi
   if (verbose)
     snprintf(ans->message[0], 500, "%s: took %.3fs\n", __func__, omp_get_wtime()-tic);
 }
+void nafillString(const SEXP *x, uint_fast64_t nx, unsigned int type, SEXP fill, ans_t *ans, bool verbose) {
+  double tic=0.0;
+  if (verbose)
+    tic = omp_get_wtime();
+  if (type==0) { // const
+    for (uint_fast64_t i=0; i<nx; i++) {
+      ((SEXP*)ans->char_v)[i] = x[i]==NA_STRING ? fill : x[i];
```

Comment:
@mattdowle can we assign to character vector (`ans->char_v`) this way? or we must use `SET_STRING_ELT`? We are assigning character that must be already in the global cache. `char_v` is `void*` that is why cast to `SEXP*` here.

Reply:
SET_STRING_ELT decrements the reference count on the old CHARSXP and increments the reference count on the CHARSXP replacing it.  So we have to use it to make sure those reference counts are maintained. Unless you know for sure that the old and new are both permanent in the global cache; e.g. NA_STRING being replaced by a `sym_` which we `install()` in init.c. But even then, it would be seen as quite risky as it would be 'behind-the-write-barrier'. Such a write may be caught by `rchk` but even if not, it's the reference counts on CHARSXP to be aware of.

Reply:
Thanks, it is now much more clear. I am trying to change code to use `SET_STRING_ELT` but I am keep getting `SET_STRING_ELT() can only be applied to a 'character vector', not a 'builtin'`. Tried redefining `ans->char_v` type and casting but haven't made it working yet. Should I expect this to be a simple change?

Reply:
@jangorecki Yes it should be simpler. If you can't work out where the `builtin` is coming from, then it could be a possible corruption. Try running the branch through the ASAN and rchk steps in CRAN_Release.cmd.  ASAN would need these lines to be covered to detect anything. Because of the reference counting, SET_STRING_ELT cannot be used inside parallel regions: try setting threads to 1 and seeing if the same error occurs to see if that might be in play.

Reply:
parallel is escaped when there is any character column on input, so that should not be an issue.

Reply:
the problem was fixed in this line
https://github.com/Rdatatable/data.table/pull/4980/commits/b24956495cb98e508fa13960c6d900ee4a6b5d1c#diff-f9176ef08eac951aa792d88fe2e2c6c3264fc1393b8836def832c33592a71932R192